### ISSUE TITLE
Setting beautiful.icon_theme = "foobarbaz" causes bad performance

### ISSUE THREAD

**Issue by psychon** [awesomeWM/awesome#1726]
This is another one of those "the menubar is slow"-issues. When no `icon_theme` is set, the code picks some defaults (I guess for most people this will be Adwaita). If an invalid theme is set... I'm not really sure what happens, but it's slow.

I ran awesome under `strace -e file` (print all file-related operations). Without changes, starting awesome, pressing mod4+p and closing results in about 2500 lines out output from strace. When setting a non-existant theme, this number goes up by a factor of 10 and the menubar becomes slow to open. Grepping for `ENOENT`, by default awesome only causes 1763 "no such entry" errors. When setting an invalid icon theme, this numbers foes up to 21906 such errors.

I guess at least some kind of warning should be printed. However, I'm not really sure which part of the code would be failing here...

(Another explanation might be: It seems by default awesome uses the Adwaita theme here, but when setting an invalid theme name, it falls back to hicolor)

**Comment by Elv13** [comment]
Same answer as always. The menubar code is based on the `awesome.freedesktop` module for Awesome 3.3-3.4. Is is doing many, many "file exists" requests and isn't very good at caching them. When it has to use the fallback paths, it gets even worst. Back in 3.4, I used that module. At the time, I rewrote half of it and it went from 94% of the syscalls to 8%. Some of these optimizations made their way over time into menubar, but not all.

Then I gave up on that code. I switched to the GDK/GTK based icon cache parser. Both Qt/KF5 and GTK use a special cache of the available icons. It is very fast and skips most "file exist"/stat attempts. Even on very slow HDD where menubar takes 10 seconds+ to load (freezing Awesome), it took < 1s. The code:

https://github.com/Elv13/awesome-configs/blob/master/utils/fd_async.lua#L411

This is the same code path used to generate Mate/GnomeShell/XFCE menu. There is no point to re-invent the wheel here. The only downside if the optional dependency on GTK. If GTK is present, then this code path should be used. Otherwise the old inefficient code path could be called. The other solution is "doing nothing". As SSD becomes used by an higher percentage of the userbase, the problem solves itself.
